### PR TITLE
Add MySQL procedure for liabilities compaction

### PR DESCRIPTION
**Pull request by shal**

### CODE REVIEW

File: db/migrate/20190719103057_add_account_recalculation_procedure.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    SQL
+  end
+
+  def up
+    drop_procedure
+
+    execute <<-SQL
+      CREATE PROCEDURE compact_and_recalculate()
+      BEGIN
+        CREATE TEMPORARY TABLE IF NOT EXISTS archive_liabilities AS (SELECT * FROM liabilities);
```
Comment (mod) [inline db/migrate/20190719103057_add_account_recalculation_procedure.rb]: I changed my mind:
CREATE DATABASE IF NOT EXISTS `peatio_archive`;
CREATE TABLE IF NOT EXISTS peatio_archive.liabilities LIKE `liabilities`;

File: db/migrate/20190719103057_add_account_recalculation_procedure.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        TRUNCATE TABLE liabilities;
+        INSERT INTO liabilities
+        (
```
Comment (mod) [inline db/migrate/20190719103057_add_account_recalculation_procedure.rb]: avoid putting (...)
This usually mean a SUB-query, which will create a temporary table, it's a heavy.
In this case I don't know what the SQL engine will do.

File: db/migrate/20190719103057_add_account_recalculation_procedure.rb [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+          SELECT NULL, code, currency_id, member_id, 'Batch', 1, SUM(debit) as debit, SUM(credit) AS credit, NOW(), NOW()
+          FROM archive_liabilities GROUP BY code, currency_id, member_id
+        );
+
+        UPDATE accounts SET
```
Comment (mod) [inline db/migrate/20190719103057_add_account_recalculation_procedure.rb]: I didn't knew this syntax, did you run this over 26 Millions records?

**Review by mod — COMMENTED** [review]
You are missing jobs table
We don't have a way to verify job run and when
also job_id will be in reference_id of the new compressed `liabilities`
Also table in archive are not temporary

File: db/migrate/20190719103057_add_account_recalculation_procedure.rb (hunk 3 above)
Comment (mod) [reply db/migrate/20190719103057_add_account_recalculation_procedure.rb]: SELECT code INTO @account_main FROM operations_accounts WHERE type = 'liability' AND kind = 'main'

UPDATE `accounts` JOIN `liabilities` AS `la` ON
 liabilities.member_id = accounts.member_id AND liabilities.currency_id = accounts.currency_id AND liabilities.code
 IN @account_main
SET accounts.balance = IFNULL(SUM(credit) - SUM(debit), 0);

File: db/migrate/20190719103057_create_jobs.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+class CreateJobs < ActiveRecord::Migration[5.2]
+  def change
+    create_table :jobs do |t|
+      t.string  :description, limit: 255
+      t.integer :rows, null: false, default: 0
```
Comment (mod) [inline db/migrate/20190719103057_create_jobs.rb]: We should add string :state to track success of the job

File: db/migrate/20190719134706_add_account_recalculation_procedure.rb [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    # Create peatio archive database.
+    execute <<-SQL
+      CREATE DATABASE IF NOT EXISTS peatio_archive;
+    SQL
+  end
+
+  def drop_archive
+    # Drop peatio archive database it was created before.
+    execute <<-SQL
+      DROP DATABASE IF NOT EXISTS peatio_archive;
```
Comment (mod) [inline db/migrate/20190719134706_add_account_recalculation_procedure.rb]: that isn't going to work, test your ROLLBACK, this syntax doesnt work

File: db/migrate/20190719134706_add_account_recalculation_procedure.rb [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    SQL
+  end
+
+  def drop_procedure
+    # Drop stored procedure if it was defined before.
+    execute <<-SQL
+      DROP PROCEDURE IF EXISTS compact_and_recalculate;
```
Comment (mod) [inline db/migrate/20190719134706_add_account_recalculation_procedure.rb]: verify this too

File: db/migrate/20190719134706_add_account_recalculation_procedure.rb [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+  end
+
+  def up
+    create_archive
+    drop_procedure
+
+    execute <<-SQL
+      CREATE PROCEDURE compact_and_recalculate()
+      BEGIN
+        DROP TABLE IF EXISTS peatio_archive.liabilities;
```
Comment (mod) [inline db/migrate/20190719134706_add_account_recalculation_procedure.rb]: the main purpose of archives is to NOT be deleted
We need a clean create table with an index on (code, currency_id, member_id, debit, credit)

File: db/migrate/20190719134706_add_account_recalculation_procedure.rb [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        CREATE TABLE peatio_archive.liabilities AS (SELECT * FROM liabilities);
+        TRUNCATE TABLE liabilities;
```
Comment (mod) [inline db/migrate/20190719134706_add_account_recalculation_procedure.rb]: space your SQL

File: db/migrate/20190719134706_add_account_recalculation_procedure.rb [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+        INSERT INTO jobs (`description`,created_at,updated_at) VALUES ('compact',NOW(),NOW());
+        INSERT INTO liabilities
+          SELECT NULL, code, currency_id, member_id, 'Job', LAST_INSERT_ID(), SUM(debit) as debit, SUM(credit) AS credit, NOW(), NOW()
+          FROM peatio_archive.liabilities GROUP BY code, currency_id, member_id;
+
+        UPDATE accounts SET
```
Comment (mod) [inline db/migrate/20190719134706_add_account_recalculation_procedure.rb]: switch to a UPDATE JOIN

File: db/migrate/20190719134706_add_account_recalculation_procedure.rb [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+            IN (SELECT code FROM operations_accounts WHERE type = 'liability' AND kind = 'main')
+        ),
+        accounts.locked =
+        (
+          SELECT IFNULL(SUM(credit) - SUM(debit), 0) FROM liabilities
+          WHERE liabilities.member_id = accounts.member_id AND liabilities.currency_id = accounts.currency_id AND liabilities.code
+            IN (SELECT code FROM operations_accounts WHERE type = 'liability' AND kind = 'locked')
+        ),
+        updated_at = NOW();
+        UPDATE jobs SET `rows` = ROW_COUNT() WHERE id = LAST_INSERT_ID();
```
Comment (mod) [inline db/migrate/20190719134706_add_account_recalculation_procedure.rb]: There is plenty of inserts above, it's unlikely LAST_INSERT_ID is correct
use: SELECT id, data INTO @x, @y FROM test.t1 LIMIT 1;
https://dev.mysql.com/doc/refman/8.0/en/select-into.html

**Review by mod — CHANGES_REQUESTED** [review]

File: db/migrate/20190719134706_add_account_recalculation_procedure.rb (hunk 7 above)
Comment (shal) [reply db/migrate/20190719134706_add_account_recalculation_procedure.rb]: Done

File: db/migrate/20190719134706_add_account_recalculation_procedure.rb (hunk 8 above)
Comment (shal) [reply db/migrate/20190719134706_add_account_recalculation_procedure.rb]: Done

File: db/migrate/20190719134706_add_account_recalculation_procedure.rb (hunk 6 above)
Comment (shal) [reply db/migrate/20190719134706_add_account_recalculation_procedure.rb]: Done

File: db/migrate/20190719134706_add_account_recalculation_procedure.rb (hunk 5 above)
Comment (shal) [reply db/migrate/20190719134706_add_account_recalculation_procedure.rb]: Done

File: db/migrate/20190719103057_create_jobs.rb (hunk 4 above)
Comment (shal) [reply db/migrate/20190719103057_create_jobs.rb]: Done

File: db/migrate/20190719134706_add_account_recalculation_procedure.rb (hunk 10 above)
Comment (shal) [reply db/migrate/20190719134706_add_account_recalculation_procedure.rb]: I did `SET @job_id = LAST_INSERT_ID();`

File: db/migrate/20190719134706_add_account_recalculation_procedure.rb (hunk 9 above)
Comment (shal) [reply db/migrate/20190719134706_add_account_recalculation_procedure.rb]: To be honest I don't clearly understand how to JOIN in this case.

**Comment by mod** [comment]
```sql
-- Declaring variables

DECLARE job_id INT DEFAULT -1;
DECLARE tmp_table VARCHAR(32);

-- Prepare job
INSERT INTO jobs (`description`, `state`, created_at, updated_at) VALUES ('compact', 'created', NOW(), NOW());

SELECT LAST_INSERT_ID() INTO job_id;
SELECT CONCAT('temp_liabilities_', job_id) INTO tmp_table;

-- Exporting raw liabilities

INSERT INTO export_liabilities
  (
    SELECT NULL, code, currency_id, member_id, reference_type, reference_id, debit, credit, created_at, updated_at
    FROM liabilities WHERE reference_type != 'Compactor'
  );

-- Compacting the ledgers
CREATE TEMPORARY TABLE tmp_table
SELECT NULL, code, currency_id, member_id, 'Compactor', job_id, SUM(debit) as debit, SUM(credit) AS credit, NOW(), NOW()
FROM liabilities GROUP BY code, currency_id, member_id;

-- Updating liabilities from temp
TRUNCATE tmp_table
INSERT INTO liabilities SELECT * FROM tmp_table;

-- Update balances

CREATE TEMPORARY TABLE `temp_balance` SELECT ac.member_id, ac.currency_id, SUM(lb.credit) - SUM(lb.debit) AS `balance` FROM `accounts` AS `ac` JOIN `liabilities` AS `lb` ON lb.code IN (SELECT code FROM operations_accounts WHERE type = 'liability' AND kind = 'main') AND ac.member_id = lb.member_id AND ac.currency_id = lb.currency_id GROUP BY lb.code, lb.currency_id, lb.member_id;

UPDATE `accounts` AS `ac` INNER JOIN `temp_balance` AS `tb` USING(member_id,currency_id)
SET ac.balance = tb.balance, `created_at` = NOW(), `updated_at` = NOW();
```

File: app/models/job.rb [hunk 11]
```diff
@@ -0,0 +1,29 @@
+# encoding: UTF-8
+# frozen_string_literal: true
+
+class Job < ApplicationRecord
+  STATES = %i[created succeed failed].freeze
+
+  validates :state, inclusion: { in: STATES }
+
+  def failed?
```
Comment (ysv) [inline app/models/job.rb]: you could try to use inquiry for state
```ruby
def state
  super&.inquiry
end
```

**Pull request closed without merging**